trace-utils/src/send_data.rs

```diff
@@ -0,0 +1,229 @@
+// Copyright 2024-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, Context};
+use futures::stream::FuturesUnordered;
+use futures::StreamExt;
+use hyper::{Body, Client, Method, Response, StatusCode};
+use std::collections::HashMap;
+
+use crate::tracer_header_tags::TracerHeaderTags;
+use datadog_trace_protobuf::pb;
+use ddcommon::{connector, Endpoint, HttpRequestBuilder};
+
+#[derive(Debug)]
+pub enum SendRequestError {
+    Hyper(hyper::Error),
+    Any(anyhow::Error),
+}
+
+pub struct SendDataResult {
+    pub last_result: anyhow::Result<Response<Body>>,
+    pub requests_count: u64,
+    pub responses_count_per_code: HashMap<u16, u64>,
+    pub errors_timeout: u64,
+    pub errors_network: u64,
+    pub errors_status_code: u64,
+}
+
+impl SendDataResult {
+    fn new() -> SendDataResult {
+        SendDataResult {
+            last_result: Err(anyhow!("No requests sent")),
+            requests_count: 0,
+            responses_count_per_code: Default::default(),
+            errors_timeout: 0,
+            errors_network: 0,
+            errors_status_code: 0,
+        }
+    }
+
+    async fn update(
```

Comment:
How do you envision this working with retry logic? Would we call `update` for each attempt of a request payload? Or should update only be called for a payload once, regardless of how many retries it took?

Reply:
Since the retry logic tries to minimize errors I'd choose the latter because it'd be confusing to send errors or dropped statistics for traces that will end up being sent.

Reply:
@hoolioh Ah, that's a great point. Maybe retry metrics should be it's own thing at some point? It may be useful information.

Reply:
I think that the retries can be deduced from the number of API requests, the number of failures of different types, as well as the number of _dropped_ traces.